lib/pam/pam.go

```diff
@@ -0,0 +1,326 @@
+// account and then opens a session.
+func Open(config *Config) (*PAM, error) {
+	if config == nil {
+		return nil, trace.BadParameter("PAM configuration is required.")
+	}
+
+	p := &PAM{
+		pamh:   nil,
+		stdin:  config.Stdin,
+		stdout: config.Stdout,
+		stderr: config.Stderr,
+	}
+
+	// Both config.ServiceName and config.Username convert between Go strings to
+	// C strings. Since the C strings are allocated on the heap in Go code, this
+	// memory must be released (and will be on the call to the Close method).
+	p.service_name = C.CString(config.ServiceName)
+	p.user = C.CString(config.Username)
+
+	// C code does not know that this PAM context exists. To ensure the
+	// conversation function can get messages to the right context, a handle
+	// registry at the package level is created (handlers). Each instance of the
+	// PAM context has it's own handle which is used to communicate between C
+	// and a instance of a PAM context.
+	p.handlerIndex = registerHandler(p)
+
+	// Create and initialize a PAM context. The pam_start function will
+	// allocate pamh if needed and the pam_end function will release any
+	// allocated memory.
+	p.retval = C._pam_start(pamHandle, p.service_name, p.user, p.conv, &p.pamh)
+	if p.retval != C.PAM_SUCCESS {
+		return nil, p.codeToError(p.retval)
+	}
+
+	// Check that the *nix account is valid. Checking a account varies based off
+	// the PAM modules used in the account stack. Typically this consists of
+	// checking if the account is expired or has access restrictions.
+	//
+	// Note: This function does not perform any authentication!
+	retval := C._pam_acct_mgmt(pamHandle, p.pamh, 0)
+	if retval != C.PAM_SUCCESS {
+		return nil, p.codeToError(retval)
+	}
+
+	// Open a user session. Opening a session varies based off the PAM modules
+	// used in the "session" stack. Opening a session typically consists of
+	// printing the MOTD, mounting a home directory, updating auth.log.
+	p.retval = C._pam_open_session(pamHandle, p.pamh, 0)
+	if p.retval != C.PAM_SUCCESS {
+		return nil, p.codeToError(p.retval)
+	}
+
+	return p, nil
+}
+
+// Close will close the session, the PAM context, and release any allocated
+// memory.
+func (p *PAM) Close() error {
+	// Close the PAM session. Closing a session can entail anything from
+	// unmounting a home directory and updating auth.log.
+	p.retval = C._pam_close_session(pamHandle, p.pamh, 0)
+	if p.retval != C.PAM_SUCCESS {
+		return p.codeToError(p.retval)
+	}
+
+	// Terminate the PAM transaction.
+	retval := C._pam_end(pamHandle, p.pamh, p.retval)
+	if retval != C.PAM_SUCCESS {
+		return p.codeToError(retval)
+	}
+
+	// Release the memory allocated for the conversation function.
+	C.free(unsafe.Pointer(p.conv))
+
+	// Release strings that were allocated when opening the PAM context.
+	C.free(unsafe.Pointer(p.service_name))
+	C.free(unsafe.Pointer(p.user))
+
+	// Unregister handler index at the package level.
+	unregisterHandler(p.handlerIndex)
```

Comment:
I think handler should be unregistered before the first free functions are called to prevent it from being called